### PR TITLE
Implement aria2 service for controlling a daemon via rpc.

### PR DESCRIPTION
**Pull request by indiscipline**
###### Motivation for this change
Implement aria2 systemd service for controlling a daemon instance via rpc. There's already a nix package for aria2, but no service unit. Aria2 can be controlled by a few freely-available web-uis via rpc calls.

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by indiscipline** [comment]
Ping @LnL7, @jgeerds, @rasendubi  (previously merged aria2-related PRs).

File: nixos/modules/misc/ids.nix [hunk 1]
```diff
@@ -547,6 +548,7 @@
       #rpc = 271; # unused
       #geoip = 272; # unused
       fcron = 273;
+	  aria2 = 274;
```
Comment (rasendubi) [inline nixos/modules/misc/ids.nix]: Indentation level.

File: nixos/modules/services/networking/aria2.nix [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+          Whether or not to enable the headless Aria2 daemon service.
+
+          Aria2 daemon can be controlled via the RPC interface using
+          one of many WebUI (http://localhost:6800/ by default).
+
+          Targets are downloaded to ${downloadDir} by default and are
+          accessible to users in the "aria2" group.
+          '';
+      };
+      openPorts = mkOption {
```
Comment (rasendubi) [inline nixos/modules/services/networking/aria2.nix]: Services should not open firewall ports. (The only exception is sshd.)

File: nixos/modules/services/networking/aria2.nix [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        type = types.string;
+        default = "aria2rpc";
+        description = ''
+          Set RPC secret authorization token.
+          Read https://aria2.github.io/manual/en/html/aria2c.html#rpc-auth to know how this option value is used.
+          '';
+      };
+      extraArguments = mkOption {
+        type = types.string;
+        example = "--rpc-secret=12345 --enable-rpc --rpc-listen-all";
```
Comment (rasendubi) [inline nixos/modules/services/networking/aria2.nix]: `rpc-secret` already has an option and `--enable-rpc` is always added, so these should not go into `extraArguments`

File: nixos/modules/services/networking/aria2.nix [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        default = false;
+        description = ''
+          Whether or not to enable the headless Aria2 daemon service.
+
+          Aria2 daemon can be controlled via the RPC interface using
+          one of many WebUI (http://localhost:6800/ by default).
+
+          Targets are downloaded to ${downloadDir} by default and are
+          accessible to users in the "aria2" group.
+          '';
```
Comment (rasendubi) [inline nixos/modules/services/networking/aria2.nix]: The usual indentation pattern for multi-line strings is next:
```nix
{
  description = ''
    blah-blah
  '';
# ^ only two spaces here
}

File: nixos/modules/services/networking/aria2.nix [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+    users.extraGroups.aria2.gid = config.ids.gids.aria2;
+
+    systemd.services.aria2 = {
+      description = "aria2 Service";
+      after = [ "local-fs.target" "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+      preStart = ''
+        mkdir -m 0770 -p "${homeDir}"
+				chown aria2:aria2 "${homeDir}"
```
Comment (rasendubi) [inline nixos/modules/services/networking/aria2.nix]: Indentation seems to be broken in `preStart`.

File: nixos/modules/services/networking/aria2.nix [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        cp -f "${settingsFile}" "${settingsDir}/aria2.conf"
+      '';
+
+      serviceConfig = {
+        Restart = "on-abort";
+        ExecStart = "${pkgs.aria2}/bin/aria2c --enable-rpc --conf-path=${settingsDir}/aria2.conf ${config.services.aria2.extraArguments} --save-session=${sessionFile}";
+        ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+        User = "aria2";
+        Group = "aria2";
+				PermissionsStartOnly = true;
```
Comment (rasendubi) [inline nixos/modules/services/networking/aria2.nix]: Indentation level.

**Review by rasendubi — COMMENTED** [review]
Superficially looks good except some minor comments.

File: nixos/modules/misc/ids.nix (hunk 1 above)
Comment (indiscipline) [inline nixos/modules/misc/ids.nix]: Fixed.

File: nixos/modules/services/networking/aria2.nix (hunk 4 above)
Comment (indiscipline) [inline nixos/modules/services/networking/aria2.nix]: Fixed.

File: nixos/modules/services/networking/aria2.nix (hunk 2 above)
Comment (indiscipline) [inline nixos/modules/services/networking/aria2.nix]: Sorry, my bad. Copied from some other service and did't give enough thought.

File: nixos/modules/services/networking/aria2.nix (hunk 3 above)
Comment (indiscipline) [inline nixos/modules/services/networking/aria2.nix]: Fixed.

File: nixos/modules/services/networking/aria2.nix (hunk 5 above)
Comment (indiscipline) [inline nixos/modules/services/networking/aria2.nix]: Fixed.

File: nixos/modules/services/networking/aria2.nix (hunk 6 above)
Comment (indiscipline) [inline nixos/modules/services/networking/aria2.nix]: Fixed.

**Comment by indiscipline** [comment]
Is there anything else to be fixed before merge?

**Comment by rasendubi** [comment]
Thanks for the contribution!

**Comment by indiscipline** [comment]
Thanks for the review, @rasendubi. All went pretty smooth for a first PR. I'll try better next time, though.

**Pull request merged**